### PR TITLE
feat: prevent rescheduling past bookings

### LINKED ISSUE #17977: [CAL-4843] Prevent rescheduling past events
When clicking on ‘reschedule‘ for past bookings it should redirect to the normal booking link and the original event won’t get canceled. 

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-4843](https://linear.app/calcom/issue/CAL-4843/prevent-rescheduling-past-events)</sub>

### PR DESCRIPTION
**Pull request by kart1ka**
## What does this PR do?

- Fixes #17977 (GitHub issue number)
- Fixes CAL-4843 (Linear issue number - should be visible at the bottom of the GitHub issue description)

https://github.com/user-attachments/assets/71b9db5b-ca6c-485e-8e6e-a6906985697f

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

### CODE REVIEW

**Review by anikdhabal — DISMISSED** [review]
Please remove the 'Reschedule' and 'Request Reschedule' options for past bookings from the dropdown

**Comment by kart1ka** [comment]
> Please remove the 'Reschedule' and 'Request Reschedule' options for past bookings from the dropdown

Done.

File: apps/web/components/booking/BookingListItem.tsx [hunk 1]
```diff
@@ -212,23 +212,27 @@ function BookingListItem(booking: BookingItemProps) {
...
+            id: "reschedule_request",
+            icon: "send" as const,
+            iconClassName: "rotate-45 w-[16px] -translate-x-0.5 ",
+            label: t("send_reschedule_request"),
+            onClick: () => {
+              setIsOpenRescheduleDialog(true);
+            },
+          },
+        ]
+      : []),
```
Comment (anikdhabal) [inline apps/web/components/booking/BookingListItem.tsx:235]: It's better to return early

Comment (kart1ka) [reply apps/web/components/booking/BookingListItem.tsx]: Done.

File: apps/web/lib/reschedule/[uid]/getServerSideProps.ts [hunk 2]
```diff
@@ -115,6 +115,28 @@ export async function getServerSideProps(context: GetServerSidePropsContext) {
...
+    ? await UserRepository.enrichUserWithItsProfile({ user: booking.user })
+    : null;
+
+  const eventUrl = await buildEventUrlFromBooking({
+    eventType,
+    dynamicGroupSlugRef: booking.dynamicGroupSlugRef ?? null,
+    profileEnrichedBookingUser: enrichedBookingUser,
+  });
+
+  const isBookingInPast = new Date(booking.endTime) < new Date();
```
Comment (youneshenniwrites) [inline apps/web/lib/reschedule/[uid]/getServerSideProps.ts:130]: What if `booking.endTime` is `null`, `undefined`, or simply invalid? Maybe it's best to add a small additional check like this:

```
const isBookingInPast = booking.endTime && new Date(booking.endTime) < new Date();
if (isBookingInPast) {
  return {
    redirect: {
      destination: eventUrl,
      permanent: false,
    },
  };
}
```

Comment (Praashh) [reply apps/web/lib/reschedule/[uid]/getServerSideProps.ts:130]: yes, it's good to have null check

**Review by Praashh — APPROVED** [review]
LGTM!!

**Comment by angus0817** [comment]
Can we revert this change? Or at least set it as an option that can be toggled on/off. The ability to reschedule past bookings was an important feature for our company. Not all clients always request an reschedule before an appointment, sometimes they forget and want to make changes at the last minute when the meeting time has already past. In this case, we are no longer able to reschedule such appointments and with this change, end up getting two appointment records for this client for a single event, when in reality they only made one.

**Pull request merged**